api/apiclient_test.go

```diff
@@ -153,14 +154,91 @@ func (s *apiclientSuite) TestServerRoot(c *gc.C) {
 }
 
 func (s *apiclientSuite) TestDialWebsocketStopped(c *gc.C) {
+	f := (*api.NewWebsocketDialer)(nil, api.DialOpts{})
 	stopped := make(chan struct{})
-	f := api.NewWebsocketDialer(nil, api.DialOpts{})
 	close(stopped)
 	result, err := f(stopped)
 	c.Assert(err, gc.Equals, parallel.ErrStopped)
 	c.Assert(result, gc.IsNil)
 }
 
+func (s *apiclientSuite) TestOpenWithSNIHostNameEmptyCert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}})
+}
+
+func (s *apiclientSuite) TestOpenWithSNIHostNameWithCACert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+		CACert:      jtesting.CACert,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}})
+}
+
+type apiDialInfo struct {
+	location   string
+	hasRootCAs bool
+	serverName string
+}
+
+// testSNIHostName tests that when the API is dialed with the given info,
+// api.newWebsocketDialer is called with the expected information
+// (one element for each call to newWebsocketDialer)
+func (s *apiclientSuite) testSNIHostName(c *gc.C, info *api.Info, expectDials []apiDialInfo) {
+	dialed := make(chan *websocket.Config)
+	s.PatchValue(api.NewWebsocketDialer, func(cfg *websocket.Config, opts api.DialOpts) func(<-chan struct{}) (io.Closer, error) {
+		return func(<-chan struct{}) (io.Closer, error) {
+			dialed <- cfg
+			return nil, errors.New("nope")
+		}
+	})
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		conn, err := api.Open(info, api.DialOpts{})
+		c.Check(conn, gc.Equals, nil)
+		c.Check(err, gc.ErrorMatches, `nope`)
+	}()
+	for _, expect := range expectDials {
```

Comment:
Ah I see now, clever, thanks! We know we can close the channel after api.Open returns because each attempt returns an error so we can be sure that they're all tried before api.Open completes.
I don't think the timeout channel is necessary at all - if things go wrong enough that the function can't send on dialed, we'll get a panic not a timeout because the channel will have been closed.

Then we end up with something like this, which is definitely much nicer, thanks again.

	func (s *apiclientSuite) testSNIHostName(c *gc.C, info *api.Info, expectDials []apiDialInfo) {
		dialed := make(chan *websocket.Config)
		fakeDialer := func(cfg *websocket.Config, opts api.DialOpts) func(<-chan struct{}) (io.Closer, error) {
			return func(<-chan struct{}) (io.Closer, error) {
				dialed <- cfg
				return nil, errors.New("nope")
			}
		}
		s.PatchValue(api.NewWebsocketDialer, fakeDialer)
		go func() {
			conn, err := api.Open(info, api.DialOpts{})
			c.Check(conn, jc.ErrorIsNil)
			c.Check(err, gc.ErrorMatches, `nope`)
			// Because all the dials return an error, we're sure that
			// all the addresses will have been tried (and the values
			// sent on dialed) before api.Open returns, so it's safe
			// to close the channel now.
			close(dialed)
		}()
	
		i := 0
		for cfg := range dialed {
			expect := expectDials[i]
			c.Check(cfg.Location.String(), gc.Equals, expect.location)
			c.Assert(cfg.TlsConfig, gc.NotNil)
			c.Check(cfg.TlsConfig.RootCAs != nil, gc.Equals, expect.hasRootCAs)
			c.Check(cfg.TlsConfig.ServerName, gc.Equals, expect.serverName)
			i++
		}
	
		c.Check(i, gc.Equals, len(expectDials))
	}